### PR TITLE
[X86][EVEX512] Restrict attaching EVEX512 for default CPU only, NFCI

### PR DESCRIPTION
**Pull request by phoebewang**
Attaching EVEX512 is used to provide backward compatibility for legacy
LLVM IR files, which didn't set EVEX512 feature explicitly.

AVX512 and AVX10 targets have set or unset EVEX512 properly through
X86.td.

However, it's not feasible to list all AVX512 and AVX10 targets or their
complementary set here to skip/restrict such code.

Instead, we can restrict it for default CPU only. "generic" is used when
"target-cpu" is not specified in IR, while "pentium4" and "x86-64" is
the default CPU if "-march" is not specified in Clang for 32-bit and
64-bit targets respectively.

This patch is no functional change intended, though it might affect
scenarios like "-march=broadwell -mavx512bw", which looks like a misuse
of "-march" and can be solved by changing to "-mtune=broadwell -mavx512bw".

### CODE REVIEW

**Comment by KanRobert** [comment]
-mavx512dw -> -mavx512bw

File: llvm/lib/Target/X86/X86Subtarget.cpp [hunk 1]
```diff
@@ -268,18 +268,23 @@ void X86Subtarget::initSubtargetFeatures(StringRef CPU, StringRef TuneCPU,
...
+    // Make sure we won't be cheated by "-avx512fp16".
+    size_t posNoAVX512F = FS.endswith("-avx512f") ? FS.size() - 8
+                                                  : FS.rfind("-avx512f,");
+    size_t posEVEX512 = FS.rfind("+evex512");
+    // Any AVX512XXX will enable AVX512F.
+    size_t posAVX512F = FS.rfind("+avx512");
+
+    if (posAVX512F != StringRef::npos &&
+        (posNoAVX512F == StringRef::npos || posNoAVX512F < posAVX512F))
+      if (posEVEX512 == StringRef::npos && posNoEVEX512 == StringRef::npos)
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86Subtarget.cpp:285]: Combine two if to one?

**Review by KanRobert — APPROVED** [review]
LGTM with two comments

File: llvm/lib/Target/X86/X86Subtarget.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86Subtarget.cpp:285]: I think keeping it separate is friendly to understand. This outer logic is for AVX512 and inner is for EVEX512.

**Comment by phoebewang** [comment]
> LGTM with two comments

Where's the second comment?

**Comment by phoebewang** [comment]
> -mavx512dw -> -mavx512bw

Oh, just saw it. Modified!

**Comment by ronlieb** [comment]
sorry for the late problem, seems our qmcpack app may have made a recent change that exposed this issue:
llc < bug.ll

llc:  llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:7536: llvm::SDValue getMemcpyLoadsAndStores(llvm::SelectionDAG&, const llvm::SDLoc&, llvm::SDValue, llvm::SDValue, llvm::SDValue, uint64_t, llvm::Align, bool, bool, llvm::MachinePointerInfo, llvm::MachinePointerInfo, const llvm::AAMDNodes&, llvm::AAResults*): Assertion `NVT.bitsGE(VT)' failed.


[repro.txt](https://github.com/llvm/llvm-project/files/13184433/repro.txt)

**Comment by phoebewang** [comment]
@ronlieb Thanks for reporting this issue! This exposed a bug not for this patch but a related one https://github.com/llvm/llvm-project/commit/24194090e17b599522a080d502ab0f68125d53dd, which should be solved by #70420.
Did you get the IR from a non Clang front end? With that change, FEs need to push a `+evex512` feature when they enable AVX512 features on a non AVX512 target (zen3 in this example) to keep the backward compatibility. Otherwise, there will be crash in the backend.

**Pull request merged**